Simplify baseline parallelization API

aewallwi opened the pull request:
Streamline the API for baseline-paralellized delay and xtalk filtering to remove lots of redundant code.

* Remove the `baseline_parallelized` scripts includeing `delay_filter_run_baseline_parallelized.py` and `xtalk_filter_run_baseline_parallelized.py`. 
* Also remove the driver functions `xtalk-filter.load_xtalk_filter_and_write_baseline_list()` and `delay_filter.load_delay_filter_and_write_baseline_list()`.
* Now `delay_filter.load_delay_filter_and_write()` and ``xtalk_filter.load_xtalk_filter_and_write()` have a `baseline_list` argument. If provided, these methods will only read in the provided baselines. They also accept `datafile_list` instead of `infilename` so out of the box they can load in lists of files and load in a user specified set of baselines.
* If a user wants to perform the time -> baseline cornerturn, they can provide a `cornerturnfile` arg to the parsers for the `xtalk_filter_run.py` and `delay_filter_run.py` scripts. If `cornerturn` file is specified, then the scripts will use its location in the `datafile_list` to determine the correct subset of baselines to load in for that job. 

These changes remove the large amount of redundant code shared between the `baseline_parallelized` versions of the scripts and also the `baseline_list` versions of the driver functions. This code redundancy also carried the potential for different drivers and scripts to diverge.

There is a small reduction in coverage because of the reduction in code (but no additional misses are introduced). Hopefully people are OK with this.

jsdillon commented on scripts/xtalk_filter_run.py:
```diff
@@ -1,20 +1,18 @@
+import hera_cal.io as io
 
-parser = xtalk_filter.xtalk_filter_argparser()
+parser = xtalk_filter.xtalk_filter_argparser(multifile=True)
```
Didn't you remove multifile?

aewallwi replied:
yeah. I swear I went through these scripts multiple times getting rid of this. I want to say there was some kind of rebase shenanigan that happened.

jsdillon commented on scripts/xtalk_filter_run.py:
```diff
@@ -23,39 +21,50 @@
+else:
+    baseline_list = None
+
+if len(baseline_list) > 0:
```
Will this break if `baseline_list` is `None?

aewallwi replied:
Yes. This existed to prevent length zero baseline lists to not throw errors, which can happen when you have more files in your data then baselines (common for two-integration files and autocorrs). I've moved the handling of this case into the covered code with some helpful warnings.

jsdillon reviewed (changes requested):
Mostly minor fixes, but looks pretty good. I like the consolidation here. 

Related, have you battle-tested your scripts? I think I've noticed a few bugs, which might be a sign that you need to move more of the code from the scripts into the module with unit tests...

aewallwi commented:
> Mostly minor fixes, but looks pretty good. I like the consolidation here.
> 
> Related, have you battle-tested your scripts? I think I've noticed a few bugs, which might be a sign that you need to move more of the code from the scripts into the module with unit tests...

Unfortunately, I haven't had a chance to do script battle testing. I expect a bit of a shake-down (with a handful of small PRs) after this all gets merged in and we start using it with the pipeline. I'll give them a hard look but the degree to which the devel branch has diverged from the main branch makes it difficult to test these.

aewallwi commented on hera_cal/xtalk_filter.py:
```diff
@@ -205,7 +137,11 @@ def load_xtalk_filter_and_write_baseline_list(datafile_list, baseline_list, calf
+        raise NotImplementedError("baseline loading and partial i/o not yet implemented.")
     hd = io.HERAData(datafile_list, filetype='uvh5', axis='blt')
+    if baseline_list is None:
+        baseline_list = hd.bls
```
Good catch. I now handle this case.

aewallwi commented on hera_cal/xtalk_filter.py:
```diff
@@ -219,43 +155,45 @@ def load_xtalk_filter_and_write_baseline_list(datafile_list, baseline_list, calf
-                           filled_outfilename=filled_outfilename, partial_write=False,
-                           clobber=clobber, add_to_history=add_to_history,
-                           extra_attrs={'Nfreqs': xf.Nfreqs, 'freq_array': np.asarray([xf.freqs])})
+    baseline_list = [bl for bl in baseline_list if bl[-1] in polarizations or len(bl) == 2]
```
I agree this isn't necessary. I should enforce that baselines are provided as a list of antpairs rather then antpairpols.

aewallwi commented on hera_cal/xtalk_filter.py:
```diff
@@ -90,94 +90,26 @@ def run_xtalk_filter(self, to_filter=None, weight_dict=None, max_frate_coeffs=[0
     Arguments:
         datafile_list: list of data files to perform cross-talk filtering on
         baseline_list: list of antenna-pair-pol triplets to filter and write out from the datafile_list.
+                       If None, load all baselines in files. Default is None.
```
The function always writes to a single file (or up to three files, one for the clean model, resid, and filled) so I don't think we need to differentiate the case when `baseline_list` is None.

aewallwi replied:
Just to make this point clearer, I've put it in the dosctring.

aewallwi commented on scripts/delay_filter_run.py:
```diff
@@ -3,14 +3,14 @@
 
-parser = delay_filter.delay_filter_argparser()
-
+parser = delay_filter.delay_filter_argparser(multifile=True)
```
yes it is.

aewallwi commented on scripts/delay_filter_run.py:
```diff
@@ -41,25 +40,34 @@
+
+baseline_list = io.baselines_from_filelist_position(filename=a.infilename, filelist=a.datafilelist)
+
+if len(baseline_list) > 0:
```
Yep. This was an errant line.

aewallwi commented on hera_cal/delay_filter.py:
```diff
@@ -201,7 +133,11 @@ def load_delay_filter_and_write_baseline_list(datafile_list, baseline_list, calf
+        raise NotImplementedError("baseline loading and partial i/o not yet implemented.")
     hd = io.HERAData(datafile_list, filetype='uvh5', axis='blt')
+    if baseline_list is None:
+        baseline_list = hd.bls
```
I've added some handling.

aewallwi commented on hera_cal/delay_filter.py:
```diff
@@ -215,34 +151,40 @@ def load_delay_filter_and_write_baseline_list(datafile_list, baseline_list, calf
-                           filled_outfilename=filled_outfilename, partial_write=False,
-                           clobber=clobber, add_to_history=add_to_history,
-                           extra_attrs={'Nfreqs': df.Nfreqs, 'freq_array': np.asarray([df.freqs])})
+    baseline_list = [bl for bl in baseline_list if bl[-1] in polarizations or len(bl) == 2]
```
I've made sure to make sure that baselines are 2-tuples and no longer have this line.

aewallwi commented on scripts/delay_filter_run.py:
```diff
@@ -27,9 +27,8 @@
     avg_red_bllens=True
-    max_contiguous_edge_flags=10000
     skip_gaps_larger_then_filter_period=False
-    skip_flagged_edges=False
```
I see it on another line in the script so I think it was a duplicate being removed.

aewallwi commented on scripts/delay_filter_run.py:
```diff
@@ -41,25 +40,34 @@
+else:
+    baseline_list = None
+
+baseline_list = io.baselines_from_filelist_position(filename=a.infilename, filelist=a.datafilelist)
```
I want to blame git rebase for this.

aewallwi commented:
Thanks for the review @jsdillon . I've gone through and addressed your comments / added some odds and ends that I missed in the reorg. I think this is ready for another look when you are ready.

jsdillon reviewed (approved):
Looks good. Thanks!
